### PR TITLE
[windows upgrade e2e] refactor to support upgrading for more new versions

### PR DESCRIPTION
**Pull request by dntczdx**
This PR refactored the code to generalize the code logic of windows_upgrade so the existing logic can be easily applied to new version. It won't be exposed to public by gcloud as a google-supported feature. Instead, it only supports the e2e test: the monthly release of windows-install-media can be verified by this tool.

1. It doesn't change any existing behaviors of 2008r2->2012r2 upgrade
2. It doesn't add any new e2e test for new versions yet.
3. It supports the new e2e test for new versions in the future.

### CODE REVIEW

**Review by bkatyl — COMMENTED** [review]
Since this changes the conditional logic on if an upgrade can occur. 

I think this needs on hold until there is a reviewed and approved design. Also to ensure the billing team agrees with the approach.

**Comment by dntczdx** [comment]
> Since this changes the conditional logic on if an upgrade can occur.
> 
> I think this needs on hold until there is a reviewed and approved design. Also to ensure the billing team agrees with the approach.

Would you tell which conditional logic is changed? I don't change any logic when I refactor it, unless I missed something.

File: cli_tools/gce_windows_upgrade/upgrade_script.ps1
Comment (bkatyl) [inline cli_tools/gce_windows_upgrade/upgrade_script.ps1:93]: Why does this use a prefix of `GCEMetadataScripts`? Shouldn't it have a prefix of the upgrade process?

File: cli_tools/gce_windows_upgrade/upgrade_script.ps1
Comment (dntczdx) [reply cli_tools/gce_windows_upgrade/upgrade_script.ps1:54]: It's not considered since previously we trust that the "license" matches the os version. As you said, to be more secure, we should exclude any unexpected version here as well. Let me add that logic.
We don't use "<" or ">" here since version may not be simply higher or lower, it can just be unexpected. For example, windows 11 vs windows server 2022.

File: cli_tools/gce_windows_upgrade/upgrade_script.ps1
Comment (dntczdx) [reply cli_tools/gce_windows_upgrade/upgrade_script.ps1:93]: The whole script is a startup script. This script will output some important status, like "GCEMetadataScripts: Beginning upgrade startup script." Our workflow needs to watch this message to ensure the machine is boot and started executing the upgrade script. We can name it with any prefix, but this prefix indicates everything happens due to the launch of GCEMetadata startup script in my opinion.

Comment (bkatyl) [reply cli_tools/gce_windows_upgrade/upgrade_script.ps1:93]: If upgrade_script.ps1 is launched using metadata startup scripts then any output from the script is captured by the GCEMetadataScripts binary and it includes the logging prefix of "GCEMetadataScripts: ".

So my concern is that the output in the serial port will look like:
"GCEMetadataScripts: GCEMetadataScripts:  post-upgrade step is done."

File: cli_tools/gce_windows_upgrade/upgrader/upgrade_paths.go
Comment (bkatyl) [inline cli_tools/gce_windows_upgrade/upgrader/upgrade_paths.go:93]: Doesn't "enabled: true" enabled upgrades for 2012r2/2016/2019?

File: cli_tools/gce_windows_upgrade/upgrader/validators.go
Comment (bkatyl) [inline cli_tools/gce_windows_upgrade/upgrader/validators.go:226]: I'd suggest a clearer error message.

No valid Windows Server PayG license can be found. Any of the following licenses are required %v"

File: cli_tools/gce_windows_upgrade/upgrader/validators.go
Comment (bkatyl) [inline cli_tools/gce_windows_upgrade/upgrader/validators.go:229]: The GCE instance has the %v license attached. This likely means the instance has already been upgraded.

Comment (bkatyl) [inline cli_tools/gce_windows_upgrade/upgrader/validators.go:229]: Can we ever have a situation where the upgrade license is attached, but the OS has failed to upgrade?

File: cli_tools/gce_windows_upgrade/upgrader/workflows.go
Comment (bkatyl) [inline cli_tools/gce_windows_upgrade/upgrader/workflows.go:183]: The order of execution for metadata startup scripts is here: https://cloud.google.com/compute/docs/instances/startup-scripts/windows#order_of_execution_of_windows_startup_scripts

Since you're using windows-startup-script-url, you'll also want to make sure that windows-startup-script-ps1, windows-startup-script-cmd, and windows-startup-script-bat don't have any values set.

File: cli_tools/gce_windows_upgrade/upgrade_script.ps1
Comment (dntczdx) [reply cli_tools/gce_windows_upgrade/upgrade_script.ps1]: You are completely right, it's duplicated. Removed it from this script.

File: cli_tools/gce_windows_upgrade/upgrader/upgrade_paths.go
Comment (dntczdx) [reply cli_tools/gce_windows_upgrade/upgrader/upgrade_paths.go:93]: In the previous commit, they should be disabled; now they all should be enabled for the e2e purpose. The gcloud gated them for the real customers, so it's not a problem to enable them for e2e.

File: cli_tools/gce_windows_upgrade/upgrader/validators.go
Comment (dntczdx) [reply cli_tools/gce_windows_upgrade/upgrader/validators.go]: done

File: cli_tools/gce_windows_upgrade/upgrader/validators.go
Comment (dntczdx) [reply cli_tools/gce_windows_upgrade/upgrader/validators.go]: yes. We have to attach the license when creating the new boot disk, which is earlier than the actual upgrade.

Comment (dntczdx) [reply cli_tools/gce_windows_upgrade/upgrader/validators.go]: adopted this words but still keeps "started an upgrade"

File: cli_tools/gce_windows_upgrade/upgrader/workflows.go
Comment (dntczdx) [reply cli_tools/gce_windows_upgrade/upgrader/workflows.go:183]: You are right, that part is missing for both manual guide and automation.
It's a bigger behavior change. Since right now my purpose is to enable the higher version upgrade only for e2e purpose (not open to customers), let's not change this behavior right now.

**Review by bkatyl — APPROVED** [review]

**Pull request merged**
